File: shared/definitions/threat_intel.py
"""Exploitation intelligence: the score feeds, the KEV catalog, and the signals derived from them."""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum


class FeedKind(StrEnum):
    EPSS = "epss"
    KEV = "kev"
    NVD = "nvd"


class FeedStatus(StrEnum):
    EMPTY = "empty"
    READY = "ready"
    STALE = "stale"
    FAILED = "failed"
    SYNCING = "syncing"


FEED_STATUS_LABELS: dict[str, str] = {
    FeedStatus.EMPTY.value: "Not downloaded",
    FeedStatus.READY.value: "Up to date",
    FeedStatus.STALE.value: "Out of date",
    FeedStatus.FAILED.value: "Last refresh failed",
    FeedStatus.SYNCING.value: "Downloading",
}


@dataclass(frozen=True)
class FeedSpec:
    kind: str
    label: str
    tagline: str
    description: str
    url: str
    source: str
    source_url: str
    license: str
    rows_table: str
    rows_noun: str


NVD_RELEASE = "https://github.com/fkie-cad/nvd-json-data-feeds/releases/latest/download"

FEEDS: tuple[FeedSpec, ...] = (
    FeedSpec(
        kind=FeedKind.EPSS.value,
        label="EPSS",
        tagline="Probability of exploitation",
        description=(
            "Modelled probability that a CVE is exploited in the wild within 30 days. "
            "Recomputed daily."
        ),
        url="https://epss.empiricalsecurity.com/epss_scores-current.csv.gz",
        source="FIRST.org",
        source_url="https://www.first.org/epss/",
        license="Free to use, no account",
        rows_table="epss_scores",
        rows_noun="scored CVEs",
    ),
    FeedSpec(
        kind=FeedKind.KEV.value,
        label="CISA KEV",
        tagline="Confirmed exploited in the wild",
        description=(
            "CVEs with confirmed exploitation in the wild, each with a remediation "
            "deadline."
        ),
        url="https://www.cisa.gov/sites/default/files/feeds/known_exploited_vulnerabilities.json",
        source="CISA",
        source_url="https://www.cisa.gov/known-exploited-vulnerabilities-catalog",
        license="Public domain (US Government)",
        rows_table="kev_entries",
        rows_noun="catalogued CVEs",
    ),
    FeedSpec(
        kind=FeedKind.NVD.value,
        label="NVD",
        tagline="Which versions a CVE affects",
        description=(
            "Every published CVE with the software versions it applies to. Matched "
            "against reported versions without sending a request."
        ),
        url=f"{NVD_RELEASE}/CVE-<year>.json.xz",
        source="NIST NVD, mirrored by fkie-cad",
        source_url="https://github.com/fkie-cad/nvd-json-data-feeds",
        license="Public domain (US Government)",
        rows_table="nvd_cves",
        rows_noun="published CVEs",
    ),
)


FEEDS_BY_KIND: dict[str, FeedSpec] = {spec.kind: spec for spec in FEEDS}

STALE_AFTER_HOURS = 48


@dataclass(frozen=True)
class ExploitBand:
    key: str
    label: str
    floor: float
    description: str


EXPLOIT_BANDS: tuple[ExploitBand, ...] = (
    ExploitBand("very_likely", "Very likely", 0.5, "EPSS 50% or above."),
    ExploitBand("likely", "Likely", 0.088, "EPSS 8.8% or above."),
    ExploitBand("possible", "Possible", 0.01, "EPSS 1% or above."),
    ExploitBand("unlikely", "Unlikely", 0.0, "EPSS below 1%."),
)

BANDS_BY_KEY: dict[str, ExploitBand] = {band.key: band for band in EXPLOIT_BANDS}
BAND_ORDER: tuple[str, ...] = tuple(band.key for band in EXPLOIT_BANDS)


def exploit_band(score: float | None) -> str | None:
    """The band a score falls in, or None when the CVE has no score."""
    if score is None:
        return None
    for band in EXPLOIT_BANDS:
        if score >= band.floor:
            return band.key
    return EXPLOIT_BANDS[-1].key


class ExploitSignal(StrEnum):
    """The reasons one finding outranks another."""

    KEV = "kev"
    RANSOMWARE = "ransomware"
    OVERDUE = "overdue"
    WEAPONISED = "weaponised"
    FRESH_EXPLOIT = "fresh_exploit"
    LIKELY = "likely"
    REACHABLE = "reachable"
    BYPASSED = "bypassed"
    RANSOM_PATH = "ransom_path"
    UNTESTABLE = "untestable"
    CROWD = "crowd"


TONE_CRITICAL = "critical"
TONE_WARNING = "warning"
TONE_INFO = "info"
TONE_NEUTRAL = "neutral"


@dataclass(frozen=True)
class SignalSpec:
    kind: str
    label: str
    help: str
    weight: int
    tone: str


SIGNALS: tuple[SignalSpec, ...] = (
    SignalSpec(
        ExploitSignal.KEV.value,
        "Known exploited",
        "CISA lists this CVE as exploited in the wild.",
        40,
        TONE_CRITICAL,
    ),
    SignalSpec(
        ExploitSignal.RANSOM_PATH.value,
        "Ransomware path",
        "Used in ransomware campaigns, and the host exposes a remote access or database service.",
        35,
        TONE_CRITICAL,
    ),
    SignalSpec(
        ExploitSignal.RANSOMWARE.value,
        "Used by ransomware",
        "CISA records this CVE in known ransomware campaigns.",
        25,
        TONE_CRITICAL,
    ),
    SignalSpec(
        ExploitSignal.FRESH_EXPLOIT.value,
        "Exploit published since the last scan",
        "A public exploit was published after the scan that recorded this finding.",
        25,
        TONE_CRITICAL,
    ),
    SignalSpec(
        ExploitSignal.OVERDUE.value,
        "Past the CISA deadline",
        "The federal remediation deadline for this CVE has passed.",
        20,
        TONE_WARNING,
    ),
    SignalSpec(
        ExploitSignal.WEAPONISED.value,
        "Public exploit available",
        "Working exploit code is published.",
        20,
        TONE_WARNING,
    ),
    SignalSpec(
        ExploitSignal.LIKELY.value,
        "Likely to be exploited",
        "EPSS 8.8% or above.",
        15,
        TONE_WARNING,
    ),
    SignalSpec(
        ExploitSignal.REACHABLE.value,
        "Directly reachable",
        "The host answers from the internet with no CDN or WAF in front of it.",
        15,
        TONE_WARNING,
    ),
    SignalSpec(
        ExploitSignal.BYPASSED.value,
        "Confirmed through a WAF",
        "A WAF or CDN sits in front of the host and the check succeeded.",
        15,
        TONE_WARNING,
    ),
    SignalSpec(
        ExploitSignal.CROWD.value,
        "Mass-scanned software",
        "Run by more than 100,000 hosts on the internet.",
        5,
        TONE_INFO,
    ),
    SignalSpec(
        ExploitSignal.UNTESTABLE.value,
        "No check exists",
        "No scanner template covers this CVE.",
        0,
        TONE_INFO,
    ),
)

SIGNALS_BY_KIND: dict[str, SignalSpec] = {spec.kind: spec for spec in SIGNALS}
SIGNAL_ORDER: tuple[str, ...] = tuple(spec.kind for spec in SIGNALS)

MAX_EXPLOIT_SCORE = 100

CROWD_HOSTS = 100_000
RANSOM_SERVICE_CLASSES: frozenset[str] = frozenset({"remote", "database"})


def exploit_score(signals: list[str] | tuple[str, ...]) -> int:
    """A finding's rank is the sum of its signal weights, capped."""
    total = sum(SIGNALS_BY_KIND[s].weight for s in signals if s in SIGNALS_BY_KIND)
    return min(total, MAX_EXPLOIT_SCORE)

File: frontend/src/lib/config/threat-intel.ts
import Biohazard from '@lucide/svelte/icons/biohazard';
import CalendarX from '@lucide/svelte/icons/calendar-x';
import EyeOff from '@lucide/svelte/icons/eye-off';
import Flame from '@lucide/svelte/icons/flame';
import Globe from '@lucide/svelte/icons/globe';
import Radiation from '@lucide/svelte/icons/radiation';
import Swords from '@lucide/svelte/icons/swords';
import TrendingUp from '@lucide/svelte/icons/trending-up';
import ShieldOff from '@lucide/svelte/icons/shield-off';
import Unplug from '@lucide/svelte/icons/unplug';
import Zap from '@lucide/svelte/icons/zap';
import type { IconComponent } from './icons';

// mirrors shared/definitions/threat_intel.py
export enum FeedKind {
	EPSS = 'epss',
	KEV = 'kev'
}

export enum FeedStatus {
	EMPTY = 'empty',
	READY = 'ready',
	STALE = 'stale',
	FAILED = 'failed',
	SYNCING = 'syncing'
}

export const FEED_STATUS_TONE: Record<string, string> = {
	[FeedStatus.READY]: 'text-success',
	[FeedStatus.STALE]: 'text-warning',
	[FeedStatus.FAILED]: 'text-destructive',
	[FeedStatus.SYNCING]: 'text-info',
	[FeedStatus.EMPTY]: 'text-muted-foreground'
};

export const FEED_STATUS_DOT: Record<string, string> = {
	[FeedStatus.READY]: 'bg-success',
	[FeedStatus.STALE]: 'bg-warning',
	[FeedStatus.FAILED]: 'bg-destructive',
	[FeedStatus.SYNCING]: 'bg-info',
	[FeedStatus.EMPTY]: 'bg-muted-foreground'
};

export enum ExploitSignal {
	KEV = 'kev',
	RANSOM_PATH = 'ransom_path',
	RANSOMWARE = 'ransomware',
	FRESH_EXPLOIT = 'fresh_exploit',
	OVERDUE = 'overdue',
	WEAPONISED = 'weaponised',
	LIKELY = 'likely',
	REACHABLE = 'reachable',
	BYPASSED = 'bypassed',
	CROWD = 'crowd',
	UNTESTABLE = 'untestable'
}

export const SIGNAL_ORDER: string[] = [
	ExploitSignal.KEV,
	ExploitSignal.RANSOM_PATH,
	ExploitSignal.RANSOMWARE,
	ExploitSignal.FRESH_EXPLOIT,
	ExploitSignal.OVERDUE,
	ExploitSignal.WEAPONISED,
	ExploitSignal.LIKELY,
	ExploitSignal.REACHABLE,
	ExploitSignal.BYPASSED,
	ExploitSignal.CROWD,
	ExploitSignal.UNTESTABLE
];

export const SIGNAL_LABELS: Record<string, string> = {
	[ExploitSignal.KEV]: 'Known exploited',
	[ExploitSignal.RANSOM_PATH]: 'Ransomware path',
	[ExploitSignal.RANSOMWARE]: 'Used by ransomware',
	[ExploitSignal.FRESH_EXPLOIT]: 'Exploit published since the last scan',
	[ExploitSignal.OVERDUE]: 'Past the CISA deadline',
	[ExploitSignal.WEAPONISED]: 'Public exploit available',
	[ExploitSignal.LIKELY]: 'Likely to be exploited',
	[ExploitSignal.REACHABLE]: 'Directly reachable',
	[ExploitSignal.BYPASSED]: 'Confirmed through a WAF',
	[ExploitSignal.CROWD]: 'Mass-scanned software',
	[ExploitSignal.UNTESTABLE]: 'No check exists'
};

export const SIGNAL_HELP: Record<string, string> = {
	[ExploitSignal.KEV]: 'CISA lists these CVEs as exploited in the wild.',
	[ExploitSignal.RANSOM_PATH]:
		'Used in ransomware campaigns. The same host exposes a sensitive service.',
	[ExploitSignal.RANSOMWARE]: 'CISA records these CVEs in known ransomware campaigns.',
	[ExploitSignal.FRESH_EXPLOIT]: 'A public exploit was published after the scan that found these.',
	[ExploitSignal.OVERDUE]: 'The federal remediation deadline for these CVEs has passed.',
	[ExploitSignal.WEAPONISED]: 'Working exploit code is published for these.',
	[ExploitSignal.LIKELY]: 'EPSS 8.8% or above.',
	[ExploitSignal.REACHABLE]:
		'The host answers from the internet with no CDN or WAF in front of it.',
	[ExploitSignal.BYPASSED]: 'A WAF or CDN sits in front of the host and the check succeeded.',
	[ExploitSignal.CROWD]: 'Hundreds of thousands of hosts run this software.',
	[ExploitSignal.UNTESTABLE]: 'No scanner template covers these.'
};

export const SIGNAL_ICONS: Record<string, IconComponent> = {
	[ExploitSignal.KEV]: Flame,
	[ExploitSignal.RANSOM_PATH]: Biohazard,
	[ExploitSignal.RANSOMWARE]: Radiation,
	[ExploitSignal.FRESH_EXPLOIT]: TrendingUp,
	[ExploitSignal.OVERDUE]: CalendarX,
	[ExploitSignal.WEAPONISED]: Swords,
	[ExploitSignal.LIKELY]: Zap,
	[ExploitSignal.REACHABLE]: Unplug,
	[ExploitSignal.BYPASSED]: ShieldOff,
	[ExploitSignal.CROWD]: Globe,
	[ExploitSignal.UNTESTABLE]: EyeOff
};

export const SIGNAL_QUERY: Record<string, string> = {
	[ExploitSignal.KEV]: 'is:kev',
	[ExploitSignal.RANSOM_PATH]: 'signal:ransom_path',
	[ExploitSignal.RANSOMWARE]: 'is:ransomware',
	[ExploitSignal.FRESH_EXPLOIT]: 'signal:fresh_exploit',
	[ExploitSignal.OVERDUE]: 'is:overdue',
	[ExploitSignal.WEAPONISED]: 'is:weaponised',
	[ExploitSignal.LIKELY]: 'signal:likely',
	[ExploitSignal.REACHABLE]: 'signal:reachable',
	[ExploitSignal.BYPASSED]: 'signal:bypassed',
	[ExploitSignal.CROWD]: 'signal:crowd',
	[ExploitSignal.UNTESTABLE]: 'is:untestable'
};

export const TONE_TEXT: Record<string, string> = {
	critical: 'text-destructive',
	warning: 'text-warning',
	info: 'text-info',
	neutral: 'text-muted-foreground'
};

export const TONE_FILL: Record<string, string> = {
	critical: 'var(--destructive)',
	warning: 'var(--warning)',
	info: 'var(--info)',
	neutral: 'var(--muted-foreground)'
};

export enum ExploitBand {
	VERY_LIKELY = 'very_likely',
	LIKELY = 'likely',
	POSSIBLE = 'possible',
	UNLIKELY = 'unlikely'
}

export const BAND_ORDER: string[] = [
	ExploitBand.VERY_LIKELY,
	ExploitBand.LIKELY,
	ExploitBand.POSSIBLE,
	ExploitBand.UNLIKELY
];

export const BAND_LABELS: Record<string, string> = {
	[ExploitBand.VERY_LIKELY]: 'Very likely',
	[ExploitBand.LIKELY]: 'Likely',
	[ExploitBand.POSSIBLE]: 'Possible',
	[ExploitBand.UNLIKELY]: 'Unlikely'
};

export const BAND_HELP: Record<string, string> = {
	[ExploitBand.VERY_LIKELY]: 'EPSS 50% or above.',
	[ExploitBand.LIKELY]: 'EPSS 8.8% or above.',
	[ExploitBand.POSSIBLE]: 'EPSS 1% or above.',
	[ExploitBand.UNLIKELY]: 'EPSS below 1%.'
};

export const BAND_FILL: Record<string, string> = {
	[ExploitBand.VERY_LIKELY]: 'var(--destructive)',
	[ExploitBand.LIKELY]: 'var(--sev-high)',
	[ExploitBand.POSSIBLE]: 'var(--warning)',
	[ExploitBand.UNLIKELY]: 'var(--muted-foreground)'
};

export const MAX_EXPLOIT_SCORE = 100;

export function bandFor(score: number | null | undefined): string | null {
	if (score === null || score === undefined) return null;
	if (score >= 0.5) return ExploitBand.VERY_LIKELY;
	if (score >= 0.088) return ExploitBand.LIKELY;
	if (score >= 0.01) return ExploitBand.POSSIBLE;
	return ExploitBand.UNLIKELY;
}

/** EPSS as a percentage. */
export function epssLabel(score: number | null | undefined): string {
	if (score === null || score === undefined) return '—';
	const pct = score * 100;
	if (pct >= 10) return `${Math.round(pct)}%`;
	if (pct >= 1) return `${pct.toFixed(1)}%`;
	if (pct >= 0.1) return `${pct.toFixed(2)}%`;
	return '<0.1%';
}

export function percentileLabel(pct: number | null | undefined): string {
	if (pct === null || pct === undefined) return '';
	return `worse than ${Math.floor(pct * 100)}% of all CVEs`;
}

export function hostsLabel(hosts: number | null | undefined): string {
	if (!hosts) return '';
	if (hosts >= 1_000_000) return `${(hosts / 1_000_000).toFixed(1)}M`;
	if (hosts >= 1_000) return `${Math.round(hosts / 1_000)}k`;
	return String(hosts);
}
